### PR TITLE
HPCC-24497 Check result size is not too big based on count

### PR DESCRIPTION
**Pull request by jakesmith**
Signed-off-by: Jake Smith <jake.smith@lexisnexisrisk.com>

## Type of change:
- [x] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [ ] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [ ] The commit message title makes sense in a changelog, by itself.
  - [ ] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

### CODE REVIEW

**Comment by jakesmith** [comment]
@wangkx @ghalliday - please review

File: esp/services/ws_workunits/ws_workunitsService.cpp [hunk 1]
```diff
@@ -2894,7 +2906,29 @@ void CWsWorkunitsEx::getWsWuResult(IEspContext &context, const char *wuid, const
...
+            result->getResultFilename(logicalName);
+        if (logicalName.length())
+        {
+            Owned<IDistributedFile> df = lookupLogicalName(context, logicalName.str());
+            if (!df)
+                throw makeStringExceptionV(ECLWATCH_FILE_NOT_EXIST, "Cannot find file %s.", logicalName.str());
+            resultSz = df->getDiskSize(true, false);
+        }
+        else
+            resultSz = result->getResultRawSize(nullptr, nullptr);
```
Comment (wangkx) [inline esp/services/ws_workunits/ws_workunitsService.cpp:2922]: Why do you prefer df->getDiskSize() more than result->getResultRawSize()? Could df->getDiskSize() return the size of whole file, but, the result is a part of the file?

**Comment by wangkx** [comment]
@jakesmith just one question

File: esp/services/ws_workunits/ws_workunitsService.cpp (hunk 1 above)
Comment (jakesmith) [reply esp/services/ws_workunits/ws_workunitsService.cpp:2922]: @wangkx - getResultRawSize will return 0, if it's a disk based result (raw result is for a workunit result only)

Comment (wangkx) [reply esp/services/ws_workunits/ws_workunitsService.cpp:2922]: I see.

**Comment by wangkx** [comment]
The code looks fine for me. @ghalliday ?

File: esp/services/ws_workunits/ws_workunitsService.cpp [hunk 2]
```diff
@@ -2844,6 +2844,22 @@ INewResultSet* createFilteredResultSet(INewResultSet* result, IArrayOf<IConstNam
...
+        return resultSz > limitSz;
+    else
+    {
+        if (start+requestCount > resultRows)
+            requestCount = resultRows-start;
+        unsigned __int64 avgRecSize = resultSz / resultRows;
+        unsigned __int64 estSize = requestCount * avgRecSize;
+        return estSize > limitSz;
+    }
+    return false;
```
[the remark this answers is not in the document]
Comment (jakesmith) [reply esp/services/ws_workunits/ws_workunitsService.cpp]: fixed

**Comment by jakesmith** [comment]
@ghalliday - removed unreachable line.

**Pull request merged**